tests/nnet/losses/test_focal_loss.py

```diff
@@ -25,76 +59,73 @@ def test_input_validation(fn, data, labels):
-    targets = target_type(targets)
-
-    probs = softmax(scores_mygrad)
-    mygrad_focal_loss = sum(truth * (-alpha * (1 - probs + 1e-14)**gamma * log(probs))) / num_datum
-    mygrad_focal_loss.backward(grad)
-
-    nn_loss = softmax_focal_loss(scores_nn, targets, alpha=alpha, gamma=gamma).mean()
-    nn_loss.backward(grad)
-
-    assert isinstance(nn_loss, Tensor) and nn_loss.ndim == 0
-    assert_allclose(nn_loss.data, mygrad_focal_loss.data, atol=1e-4, rtol=1e-4)
-    assert_allclose(scores_nn.grad, scores_mygrad.grad, atol=1e-4, rtol=1e-4)
+@given(gamma=st.floats(max_value=0, exclude_max=True))
+def test_raises_on_bad_gamma(gamma: float):
+    with pytest.raises(ValueError):
+        focal_loss(np.array([[1.0]]), np.array([0]), gamma=gamma)
 
-    nn_loss.null_gradients()
-    assert scores_nn.grad is None
 
-
-@given(
-    num_datum=st.integers(1, 100),
-    num_classes=st.integers(1, 15),
-    alpha=st.floats(-1, 1),
-    gamma=st.floats(0, 5),
-    data=st.data(),
-    grad=st.floats(-1, 1),
-    target_type=st.sampled_from((np.array, Tensor)),
+@fwdprop_test_factory(
+    mygrad_func=focal_loss,
+    true_func=numpy_focal_loss,
+    index_to_arr_shapes={0: hnp.array_shapes(min_dims=2, max_dims=2)},
+    index_to_bnds={0: (1e-14, 100)},
+    num_arrays=1,
+    kwargs=dict(
+        targets=lambda scores: targets(scores=scores),
+        alpha=lambda scores: st.floats(-2, 2),
+        gamma=lambda scores: st.floats(0, 10),
+    ),
 )
-def test_focal_loss(num_datum, num_classes, alpha, gamma, data, grad, target_type):
-    scores = data.draw(
-        hnp.arrays(shape=(num_datum, num_classes), dtype=float, elements=st.floats(1, 100))
-    )
-    assume((abs(scores.sum(axis=1)) > 0.001).all())
-
-    scores_mygrad = Tensor(scores)
-    scores_nn = Tensor(scores)
-
-    truth = np.zeros((num_datum, num_classes))
-    targets = data.draw(st.tuples(*(st.integers(0, num_classes - 1) for i in range(num_datum))))
-    truth[range(num_datum), targets] = 1
-    targets = target_type(targets)
-
-    fl = focal_loss(softmax(scores_mygrad), targets, alpha=alpha, gamma=gamma).mean()
-    fl.backward(grad)
-
-    nn_loss = softmax_focal_loss(scores_nn, targets, alpha=alpha, gamma=gamma).mean()
-    nn_loss.backward(grad)
-
-    assert isinstance(nn_loss, Tensor) and nn_loss.ndim == 0
-    assert_allclose(nn_loss.data, fl.data, atol=1e-4, rtol=1e-4)
-    assert_allclose(scores_nn.grad, scores_mygrad.grad, atol=1e-4, rtol=1e-4)
-
-    nn_loss.null_gradients()
-    assert scores_nn.grad is None
+def test_focal_fwd():
+    pass
+
+
+@backprop_test_factory(
+    mygrad_func=focal_loss,
+    true_func=numpy_focal_loss,
+    index_to_arr_shapes={0: hnp.array_shapes(min_dims=2, max_dims=2)},
+    index_to_bnds={0: (1e-14, 1)},
+    num_arrays=1,
+    kwargs=dict(
+        targets=lambda scores: targets(scores=scores),
+        alpha=lambda scores: st.floats(-2, 2),
+        gamma=lambda scores: st.floats(0, 10) | st.sampled_from([0.0, 1.0]),
```

Comment:
I'm assuming the `| st.sampled_from([0, 1])` is just to increase the chance of those exact values being exercised? Is it possible to provide them as `@example`s?

Reply:
Yeah that just helps to ensure those edge cases get exercises.

I wish there was a way to do "partial" examples like this, but there isn't. I believe you have to provide examples for everything in the signature.

I'll have to check with the hypothesis devs to see if they are interested in changing this.